src/day_three/mod.rs

```diff
@@ -0,0 +1,129 @@
+// https://adventofcode.com/2021/day/3
+pub fn execute() {
+    let input = include_str!("input.txt");
+
+    let mut num_bits = 0;
+    let sensor_readings: Vec<u16> = input
+        .lines()
+        .filter_map(|s| match u16::from_str_radix(s, 2) {
+            Ok(i) => {
+                if s.trim().len() > num_bits {
+                    num_bits = s.trim().len()
+                }
+
+                return Option::from(i);
+            }
+            Err(_) => None,
+        })
+        .collect();
+
+    println!(
+        "Power consumption: {}",
+        power_consumption(&sensor_readings, num_bits)
+    );
+    println!(
+        "Life Support Rating: {}",
+        life_support_rating(&sensor_readings, num_bits)
+    );
+}
+
+fn power_consumption(sensor_readings: &[u16], num_bits: usize) -> u32 {
+    let bit_counts: Vec<u32> = count_flagged_bits_by_position(sensor_readings, num_bits);
+
+    let mut gamma_rate: u32 = 0;
+    let mut epsilon_rate: u32 = 0;
+
+    for i in 0..num_bits {
+        let mask = 1 << i;
+
+        // If this bit is 1 for more than half the readings, 1 is most significant
+        if bit_counts[i] > sensor_readings.len() as u32 - bit_counts[i] {
+            gamma_rate |= mask
+        } else {
+            epsilon_rate |= mask
+        }
+    }
+
+    return gamma_rate * epsilon_rate;
+}
+
+fn filter_candidates<RetainCriteria: Fn(u32, u32) -> bool>(
```

Comment:
AFAIK there isn't a way to give rust closure arguments names, which is surprising as far as missing features go IMO, considering how prevalent they are in the stdlib

Reply:
Can't say I'm too surprised. `std::function` in C++ doesn't support named arguments either and it's also used a lot in the stdlib.